{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      ">>> Oligos design for OCA2_8\n",
      "\n",
      "Annealing view\n",
      "5'- CACCGAATAAAAAGTAAGTCCAGGC -3'\n",
      "        *||||||||||||||||||||\n",
      "    3'- CTTATTTTTCATTCAGGTCCGCAAA -5'\n",
      "\n",
      "Oligos to order (5'-3' orientation)\n",
      "OCA2_8_Top    : CACCGAATAAAAAGTAAGTCCAGGC\n",
      "OCA2_8_Bottom : AAACGCCTGGACTTACTTTTTATTC\n",
      "\n",
      "\n",
      ">>> Oligos design for OCA2_1\n",
      "\n",
      "Annealing view\n",
      "5'- CACCGCACTGAAGCCTCGACTTTCC -3'\n",
      "        *||||||||||||||||||||\n",
      "    3'- CGTGACTTCGGAGCTGAAAGGCAAA -5'\n",
      "\n",
      "Oligos to order (5'-3' orientation)\n",
      "OCA2_1_Top    : CACCGCACTGAAGCCTCGACTTTCC\n",
      "OCA2_1_Bottom : AAACGGAAAGTCGAGGCTTCAGTGC\n",
      "\n",
      "\n",
      ">>> Oligos design for OCA2_2\n",
      "\n",
      "Annealing view\n",
      "5'- CACCGAGTCGAGGCTTCAGTGAGTC -3'\n",
      "        *||||||||||||||||||||\n",
      "    3'- CTCAGCTCCGAAGTCACTCAGCAAA -5'\n",
      "\n",
      "Oligos to order (5'-3' orientation)\n",
      "OCA2_2_Top    : CACCGAGTCGAGGCTTCAGTGAGTC\n",
      "OCA2_2_Bottom : AAACGACTCACTGAAGCCTCGACTC\n",
      "\n",
      "\n",
      ">>> Oligos design for HERC2_9\n",
      "\n",
      "Annealing view\n",
      "5'- CACCGATATATAAAATTTAATCAAC -3'\n",
      "        *||||||||||||||||||||\n",
      "    3'- CTATATATTTTAAATTAGTTGCAAA -5'\n",
      "\n",
      "Oligos to order (5'-3' orientation)\n",
      "HERC2_9_Top    : CACCGATATATAAAATTTAATCAAC\n",
      "HERC2_9_Bottom : AAACGTTGATTAAATTTTATATATC\n",
      "\n",
      "\n",
      ">>> Oligos design for HERC2_1\n",
      "\n",
      "Annealing view\n",
      "5'- CACCGGGTGGGAGGTTCAGATTGAA -3'\n",
      "        *||||||||||||||||||||\n",
      "    3'- CCCACCCTCCAAGTCTAACTTCAAA -5'\n",
      "\n",
      "Oligos to order (5'-3' orientation)\n",
      "HERC2_1_Top    : CACCGGGTGGGAGGTTCAGATTGAA\n",
      "HERC2_1_Bottom : AAACTTCAATCTGAACCTCCCACCC\n",
      "\n",
      "\n",
      ">>> Oligos design for RE_Ctrl FAILED !!!\n",
      "##Warning!! XbaI restriction site found! in ('RE_Ctrl', 'AGTCTGTCTAGACTCGAAAA')\n",
      "\n",
      ">>> Oligos design for HERC2_10\n",
      "\n",
      "Annealing view\n",
      "5'- CACCGTGATTAAATTTTATATATGG -3'\n",
      "        *||||||||||||||||||||\n",
      "    3'- CACTAATTTAAAATATATACCCAAA -5'\n",
      "\n",
      "Oligos to order (5'-3' orientation)\n",
      "HERC2_10_Top    : CACCGTGATTAAATTTTATATATGG\n",
      "HERC2_10_Bottom : AAACCCATATATAAAATTTAATCAC\n",
      "\n",
      "#################################################\n",
      "##WARNING! One or more sequences are incorrect.##\n",
      "#################################################\n",
      "Please check the following sequence(s):\n",
      "[('RE_Ctrl', 'AGTCTGTCTAGACTCGAAAA')]\n"
     ]
    }
   ],
   "source": [
    "#This script takes a \"Guide sequence\" generated by CRISPR Design Tool (http://tools.genome-engineering.org)\n",
    "#and generates the oligos needed for their cloning into a pX-series vector.\n",
    "#For more info check: Ran, F. A. et al. Genome engineering using the CRISPR-Cas9 system. Nat Protoc 8, 2281–2308 (2013).\n",
    "\n",
    "#Input variable: sequence_list - It's a list of tuples, [('seq_name','sequence'),...].\n",
    "#Each tuple contains the name of the sequence and the sequence.\n",
    "\n",
    "#author: alessio.marcozzi@gmail.com\n",
    "#version: 1.4\n",
    "#date last revision: January 2016\n",
    "\n",
    "\n",
    "##Insert your input sequence(s) here and hit 'Play'.\n",
    "##Do not include the PAM tor your input sequence(s)!\n",
    "sequence_list = [('OCA2_8','AATAAAAAGTAAGTCCAGGC'),\n",
    "                 ('OCA2_1','CACTGAAGCCTCGACTTTCC'),\n",
    "                 ('OCA2_2','AGTCGAGGCTTCAGTGAGTC'),\n",
    "                 ('HERC2_9','ATATATAAAATTTAATCAAC'),\n",
    "                 ('HERC2_1','GGTGGGAGGTTCAGATTGAA'),\n",
    "                 ('RE_Ctrl','AGTCTGTCTAGACTCGAAAA'),\n",
    "                 ('HERC2_10','TGATTAAATTTTATATATGG')]\n",
    "\n",
    "\n",
    "##Global variables\n",
    "BAD_SEQUENCES = []\n",
    "\n",
    "\n",
    "##Functions declaration\n",
    "def check(sequence):\n",
    "    '''Checks a DNA sequence.\n",
    "    Returns False if something inappropriate is found.'''\n",
    "    global BAD_SEQUENCES\n",
    "    #check for non ATCG letters\n",
    "    for letter in sequence:\n",
    "        if letter.upper() not in 'ATCG':\n",
    "            print(\"\\n>>> Oligos design for {} FAILED !!!\".format(guide_name))\n",
    "            print(\"##Warning!! Non ATCG base found! in {}\".format((guide_name, sequence)))\n",
    "            BAD_SEQUENCES.append((guide_name,sequence))\n",
    "            return False\n",
    "    #check for the presence of unwanted restriction sites\n",
    "    re_sites = {'BbsI_F':'GAAGAC',\n",
    "                'BbsI_R':'GTCTTC',\n",
    "                'XbaI':'TCTAGA',\n",
    "                'KpnI':'GGTACC'}\n",
    "    for k, v in re_sites.items():\n",
    "        if v in sequence:\n",
    "            print(\"\\n>>> Oligos design for {} FAILED !!!\".format(guide_name))\n",
    "            print(\"##Warning!! {} restriction site found! in {}\".format(k, (guide_name, sequence)))\n",
    "            BAD_SEQUENCES.append((guide_name, sequence))\n",
    "            return False\n",
    "    return True      \n",
    "\n",
    "def reverse(sequence):\n",
    "    '''Takes a DNA sequence as input.\n",
    "    Returns its reverse sequence.'''\n",
    "    reversed_sequence = ''\n",
    "    for n in range(len(sequence)):\n",
    "        reversed_sequence += sequence[-(1+n)]\n",
    "    return reversed_sequence\n",
    "\n",
    "def complement(sequence):\n",
    "    '''Takes a DNA sequence as input.\n",
    "    Returns its complementary sequence.'''\n",
    "    complementary_sequence = ''\n",
    "    for base in sequence:\n",
    "        if base == 'A':\n",
    "            complementary_sequence += 'T'\n",
    "        elif base == 'T':\n",
    "            complementary_sequence += 'A'\n",
    "        elif base == 'C':\n",
    "            complementary_sequence += 'G'\n",
    "        elif base == 'G':\n",
    "            complementary_sequence += 'C'\n",
    "        else:\n",
    "            raise ValueError(\"##Error!! Non ATCG base found! in {}\".format(sequence))\n",
    "    return complementary_sequence\n",
    "\n",
    "def output(sequence, guide_name):\n",
    "    '''Prints a human readable output.'''\n",
    "    forward_extra = 'CACCG' # 5'-CACCG-3' BbsI + G\n",
    "    reverse_extra = 'CAAA' # 3'-CAAA-5' BbsI\n",
    "    fd = forward_extra + sequence\n",
    "    rv = complement('G' + sequence) + reverse_extra\n",
    "    print(\"\\n>>> Oligos design for {}\\n\".format(guide_name))\n",
    "    print(\"Annealing view\")\n",
    "    print(\"5'-\", fd, \"-3'\")\n",
    "    print(\"       \", \"*||||||||||||||||||||\") # '*' indicates the extra G \"An additional cytosine\n",
    "                                              # As specified in Genome engineering using the CRISPR-Cas9 system.\n",
    "                                              # Nature Protocols, 8(11), 2281–2308.\n",
    "                                              # http://doi.org/10.1038/nprot.2013.143\n",
    "                                              # (‘C’ in gray rectangle) is appended in the reverse primer\n",
    "                                              # directly 3′ to the target sequence to allow guanine\n",
    "                                              # as the first base of the U6 transcript.\n",
    "    print(\"   \", \"3'-\", rv, \"-5'\")\n",
    "    print(\"\\nOligos to order (5'-3' orientation)\")\n",
    "    print(\"{}_Top    : {}\".format(guide_name,fd))\n",
    "    print(\"{}_Bottom : {}\\n\".format(guide_name,reverse(rv)))\n",
    "\n",
    "#print('Done')\n",
    "\n",
    "\n",
    "##Body\n",
    "for item in sequence_list:\n",
    "    guide_name = item[0]\n",
    "    sequence = item[1]\n",
    "    if check(sequence) == True:\n",
    "        output(sequence, guide_name)\n",
    "    if len(sequence) > 20:\n",
    "        print('Sequence {}:{} is longer than expected'.format(sequence,guide_name))\n",
    "        BAD_SEQUENCES.append((guide_name,sequence))\n",
    "        \n",
    "\n",
    "##Errors tracking\n",
    "if len(BAD_SEQUENCES) > 0:\n",
    "    print('#################################################')\n",
    "    print(\"##WARNING! One or more sequences are incorrect.##\")\n",
    "    print('#################################################')\n",
    "    print(\"Please check the following sequence(s):\")\n",
    "    print(BAD_SEQUENCES)\n",
    "else:\n",
    "    print(\"All the guides were correctly designed!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
